public/app/features/dashboard/api/dashboard_api.ts

```diff
@@ -1,147 +1,52 @@
+export function setDashboardAPI(override: Partial<DashboardAPIClients> | undefined) {
+  if (process.env.NODE_ENV !== 'test') {
+    throw new Error('dashboardAPI can be only overridden in test environment');
   }
+  clients = override;
 }
 
-export interface DashboardWithAccessInfo<T = DashboardDataDTO> extends Resource<T, 'DashboardWithAccessInfo'> {
-  access: Object; // TODO...
-}
-
-// Implemented using /apis/dashboard.grafana.app/*
-class K8sDashboardAPI implements DashboardAPI {
-  private client: ResourceClient<DashboardDataDTO>;
-
-  constructor() {
-    this.client = new ScopedResourceClient<DashboardDataDTO>({
-      group: 'dashboard.grafana.app',
-      version: 'v0alpha1',
-      resource: 'dashboards',
-    });
-  }
-
-  saveDashboard(options: SaveDashboardCommand): Promise<SaveDashboardResponseDTO> {
-    const dashboard = options.dashboard as DashboardDataDTO; // type for the uid property
-    const obj: ResourceForCreate<DashboardDataDTO> = {
-      metadata: {
-        ...options?.k8s,
-      },
-      spec: {
-        ...dashboard,
-      },
-    };
-
-    if (options.message) {
-      obj.metadata.annotations = {
-        ...obj.metadata.annotations,
-        [AnnoKeyMessage]: options.message,
-      };
-    } else if (obj.metadata.annotations) {
-      delete obj.metadata.annotations[AnnoKeyMessage];
-    }
-
-    if (options.folderUid) {
-      obj.metadata.annotations = {
-        ...obj.metadata.annotations,
-        [AnnoKeyFolder]: options.folderUid,
-      };
-    }
-
-    if (dashboard.uid) {
-      obj.metadata.name = dashboard.uid;
-      return this.client.update(obj).then((v) => this.asSaveDashboardResponseDTO(v));
-    }
-    return this.client.create(obj).then((v) => this.asSaveDashboardResponseDTO(v));
-  }
-
-  asSaveDashboardResponseDTO(v: Resource<DashboardDataDTO>): SaveDashboardResponseDTO {
-    return {
-      uid: v.metadata.name,
-      version: v.spec.version ?? 0,
-      id: v.spec.id ?? 0,
-      status: 'success',
-      slug: '',
-      url: '',
+// Overloads
+export function getDashboardAPI(): DashboardAPI<DashboardDTO>;
+export function getDashboardAPI(requestV2Response: 'v2'): DashboardAPI<DashboardWithAccessInfo<DashboardV2Spec>>;
+export function getDashboardAPI(requestV2Response?: 'v2'): DashboardAPI<DashboardReturnTypes> {
+  const v = getDashboardsApiVersion();
+  const isConvertingToV1 = !requestV2Response;
+
+  if (!clients) {
+    clients = {
+      legacy: new LegacyDashboardAPI(),
+      v0: new K8sDashboardAPI(),
+      v2: new K8sDashboardV2APIStub(isConvertingToV1),
     };
   }
 
-  deleteDashboard(uid: string, showSuccessAlert: boolean): Promise<DeleteDashboardResponse> {
-    return this.client.delete(uid).then((v) => ({
-      id: 0,
-      message: v.message,
-      title: 'deleted',
-    }));
+  if (v === 'v2' && requestV2Response === 'v2') {
+    return new K8sDashboardV2APIStub(false);
   }
 
```

Comment:
@axelavargas  - this make it non-testable using `setDashboardAPI` in the test files. Any suggestions how we can allow that? Also it's weird we create new instance every time v2 is requested.

Reply:
ignore actually, we can mock the responses in a simpler way that with `setDashboardAPI` :)